_adr/92/index.adoc

```diff
@@ -0,0 +1,105 @@
+
+* the services are exposing APIs for the offered specific functionality, but we lack an integration layer
+* the data aggregation logic is duplicated, scattered, and eventually diverging in various codebases (especially UI and CLI)
+* there is no "single source of truth" for data aggregated from various services
+* the current REST API layer exposes the user to the entire complexity of the underlying architecture
+
+We want to deduplicate the development effort and provide a more cohesive UX to the users.
+
+## Goals
+
+* a more cohesive UX experience across UI and CLI
+* flow simplification for the REST API exposing opinionated endpoints directly mapping UI/CLI needs
+* a single place with shared logic for aggregating and serving data
+* reduce the interactions over the network for providing a given functionality
+* easy testing/mocking setup for UI/CLI
+
+## Non-goals
+
+* provide a generic API layer for consuming the current services APIs
+* provide a generic queryable layer for the current services APIs
+
+## Current situation
+
+In the current architecture both UI and CLI are reaching out both Control Plane and Data Plane REST API endpoints.
+
+image::current_architecture.png[Current Architecture]
+
+
+## Proposal
+
+[NOTE]
+*Assumption* as a first iteration we are assuming that the CLI needs can be covered by a strict subset of the functionalities needed for the UI.
+
+We propose to integrate *one* server side component commonly known as https://samnewman.io/patterns/architectural/bff/[BFF] (Backend For Frontend).
+
+image::with_BFF.png[Proposed Architecture]
+
+The service is implemented using Java and Quarkus to achieve maximum collaboration across the current teams and align with Red Hat tech strategy.
+The API layer is designed using API Designer(RHOAD) and conforming to the OpenAPI 3+ specification to achieve maximum collaboration across different teams and members with various technological background.
+The development is driven by a "Contract First" approach leveraging the [Apicurio Codegen Quarkus Extension](https://github.com/Apicurio/apicurio-codegen/tree/main/quarkus-extension) and re-using the Java SDKs(based on Kiota) provided by the DevExp Team.
+The BFF have to to be stateless so that it can easily be scaled up and down according to the needs.
+All of the exposed endpoints are going to perform 1+ REST APIs calls to the underlying services and provide back the result.
+
+### Threat model
+
+Being it a publicly accessible service the generic service Threat model applies, we should ensure stability of the platform by having all the good practices in place:
+
+* Rate Limiting
+* Circuit Breaking
+* Caching
+
+Regarding multitenancy, it can be achieved, in the first prototype by re-using the Access Token for the initial API request to issue the following requests to the backend services.
+
+## Alternatives Considered / Rejected
+
+### GraphQL
+
```

Comment:
Given GraphQL is the best known standardized architecture for BFF's, I think it's rejection deserves a well thought out and reasoned argument.

Reply:
> Given GraphQL is the best known standardized architecture for BFF's

Curious to look at the numbers/sources, can you please share?

> I think it's rejection deserves a well thought out and reasoned argument.

I'll review this section by the EOD with new input from a chat with @machi1990 and @EricWittmann

Reply:
Don't have any numbers off hand.  Maybe someone can go to this years GraphQL conference and gather intel.  Or maybe find the info in some of last year's sessions: https://www.apollographql.com/events/virtual-event/graphql-summit/#on-demand

Reply:
Honestly interested in a more "independent" data source other than a conference on the technology!

E.g. I love the Scala programming language but I won't go to a Scala conference to gather data about "how much it spreads", there is an obvious bias there.

Reply:
It is a great way to judge community and learn from other folks implementing it.  If the conference is small, then not to many folks use it or are interested in it.  Like what ever we come up with would have ZERO ecosystem in comparison.